src/test/java/com/segment/analytics/android/integration/firebase/FirebaseTest.java

```diff
@@ -1,5 +1,148 @@
+        verify(firebase).setUserId("foo");
+    }
+
+    @Test
+    public void identifyWithTraits() {
+        Traits traits = createTraits("foo")
+                .putAge(20)
+                .putFirstName("bar")
+                .putLastName("baz")
+                .putValue("anonymousId", 123)
+                .putValue("Sign Up Date", new Date())
+                .putValue("  extra spaces        ", "bar");
+
+
+        integration.identify(new IdentifyPayloadBuilder().traits(traits).build());
+
+        verify(firebase).setUserId("foo");
+        verify(firebase).setUserProperty("firstName", "bar");
+        verify(firebase).setUserProperty("lastName", "baz");
+        verify(firebase).setUserProperty("anonymousId", "123");
+        verify(firebase).setUserProperty("Sign_Up_Date", "2017-09-07");
+        verify(firebase).setUserProperty("extra_spaces", "bar");
+    }
+
+    @Test
+    public void track() {
+        integration.track(new TrackPayloadBuilder().event("foo").build());
+        verify(firebase).logEvent(eq("foo"), bundleEq(new Bundle()));
+    }
+
+    @Test
+    public void trackWithProperties() {
+        Properties properties = new Properties()
+                .putValue("integer", 1)
+                .putValue("double", 1.0)
+                .putValue("string", "foo")
+                .putValue("date", new Date(117, 0, 1))
+                .putValue("key with spaces", "bar")
+                .putValue("total", 100.0)
+                .putValue("  extra spaces   ", "baz");
+
+        integration.track(new TrackPayloadBuilder().properties(properties).event("foo").build());
+
+        Bundle expected = new Bundle();
+        expected.putInt("integer", 1);
+        expected.putDouble("double", 1.0);
+        expected.putString("string", "foo");
+        expected.putString("date", "2017-01-01");
+        expected.putString("key_with_spaces", "bar");
+        expected.putDouble("value", 100.0);
+        expected.putString("currency", "USD");
+        expected.putString("extra_spaces", "baz");
+
+        verify(firebase).logEvent(eq("foo"), bundleEq(expected));
+    }
+
+    @Test
+    public void testMakeKey() {
+        String string = " long event name with random spaces that is more than forty characters long           ";
+        assertThat(integration.makeKey(string)).isEqualTo("long_event_name_with_random_spaces_that_");
+    }
+
+    public static Bundle bundleEq(Bundle expected) {
+        return argThat(new BundleObjectMatcher(expected));
+    }
+
+    private static class BundleObjectMatcher extends TypeSafeMatcher<Bundle> {
+        private final Bundle expected;
+
+        private BundleObjectMatcher(Bundle expected) {
+            this.expected = expected;
+        }
+
+        @Override public boolean matchesSafely(Bundle bundleObject) {
+            // todo: this relies on having the same order
+            return expected.toString().equals(bundleObject.toString());
```

Comment:
ah this isn't going to work. The example I gave just works for JSONObject. You need to add your own code to compare the two bundle objects and return true/false depending on if they match.

Reply:
The tests pass currently. Are these false positives?

Reply:
Also, I played around with values that I knew would cause the tests to fail. The tests pass/fail as expected given the values I'm setting.

Reply:
Oh interesting. It looks like they do have a custom toString implementation http://grepcode.com/file/repo1.maven.org/maven2/org.robolectric/android-all/4.3_r2-robolectric-0/android/os/Bundle.java#Bundle.toString%28%29, http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b27/java/util/AbstractMap.java#AbstractMap.toString%28%29.

So this should be ok for now!

Reply:
Thank goodness! :)